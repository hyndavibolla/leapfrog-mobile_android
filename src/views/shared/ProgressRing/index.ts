export { default as ProgressRing } from './ProgressRing';
