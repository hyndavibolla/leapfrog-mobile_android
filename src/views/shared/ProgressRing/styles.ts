import { css } from '@emotion/native';

import { COLOR } from '_constants';

export const styles = {
  container: css`
    flex: 1;
    align-items: center;
  `,
  progressRing: css`
    border-radius: 8px;
    background-color: ${COLOR.WHITE};
    transform: rotateZ(270deg);
  `
};
