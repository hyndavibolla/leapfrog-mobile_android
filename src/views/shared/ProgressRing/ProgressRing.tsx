import React, { memo } from 'react';
import Svg, { Circle, Text } from 'react-native-svg';
import { View, StyleProp, ViewStyle } from 'react-native';

import { styles } from './styles';
import { COLOR, FONT_FAMILY } from '../../../constants';

export interface Props {
  width?: number;
  strokeWidth?: number;
  color?: COLOR;
  textColor?: COLOR;
  progress: number;
  label?: string;
  sublabel?: string;
  style?: StyleProp<ViewStyle>;
}

export const ProgressRing = ({
  width = 175,
  strokeWidth = 15,
  color = COLOR.PRIMARY_LIGHT_BLUE,
  textColor = COLOR.PRIMARY_BLUE,
  progress,
  label = '',
  sublabel = '',
  style
}: Props) => {
  const size = width - 32;
  const progressStrokeWidth = strokeWidth;
  const placeholderStrokeWidth = strokeWidth + 1;
  const radius = (size - progressStrokeWidth) / 2;
  const circumference = radius * 2 * Math.PI;
  const strokeDashoffset = ((progress * circumference) / 100) * -1;
  const center = size / 2;
  const sublabelAxisYOffset = center + 15;
  const zeroPointRadius = strokeWidth / 2;

  return (
    <View style={[styles.container, style]}>
      <Svg width={size} height={size} style={styles.progressRing}>
        <Circle stroke={color} fill="none" cx={center} cy={center} r={radius} strokeWidth={progressStrokeWidth} />
        <Circle
          stroke={COLOR.MEDIUM_GRAY}
          fill="none"
          cx={size / 2}
          cy={size / 2}
          r={radius}
          strokeDasharray={`${circumference} ${circumference}`}
          strokeWidth={placeholderStrokeWidth}
          strokeDashoffset={strokeDashoffset}
        />
        <Text
          x={center}
          y={center}
          fontFamily={FONT_FAMILY.BOLD}
          textAnchor="middle"
          fill={textColor}
          stroke={textColor}
          fontSize="18px"
          transform={`rotate(90, ${center}, ${center})`}
          letterSpacing={3}
        >
          {label}
        </Text>
        <Text
          x={center}
          y={sublabelAxisYOffset}
          fontFamily={FONT_FAMILY.BOLD}
          fontWeight="bold"
          textAnchor="middle"
          fill={COLOR.DARK_GRAY}
          fontSize="12px"
          transform={`rotate(90, ${center}, ${center})`}
        >
          {sublabel}
        </Text>
        {progress === 0 ? <Circle fill={color} cx={center + radius} cy={center} r={zeroPointRadius} /> : null}
      </Svg>
    </View>
  );
};

export default memo(ProgressRing);
