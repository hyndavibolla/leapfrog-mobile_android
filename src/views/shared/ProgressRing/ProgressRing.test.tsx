import React from 'react';
import { render } from '@testing-library/react-native';

import ProgressRing, { Props } from './ProgressRing';
import { COLOR } from '_constants';

describe('ProgressRing', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      progress: 30,
      label: '31,500',
      sublabel: 'OF 34,999'
    };
  });

  it('should render ', () => {
    const { toJSON } = render(<ProgressRing {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with specific width ', () => {
    const { toJSON } = render(<ProgressRing {...props} width={175} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with specific strokeWidth', () => {
    const { toJSON } = render(<ProgressRing {...props} strokeWidth={15} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with specific color', () => {
    const { toJSON } = render(<ProgressRing {...props} color={COLOR.PRIMARY_LIGHT_BLUE} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with zero progress', () => {
    const { toJSON } = render(<ProgressRing {...props} progress={0} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
