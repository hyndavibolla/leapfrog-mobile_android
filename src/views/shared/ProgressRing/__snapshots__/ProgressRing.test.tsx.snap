// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressRing should render  1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    bbHeight={143}
    bbWidth={143}
    focusable={false}
    height={143}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "transform": Array [
            Object {
              "rotateZ": "270deg",
            },
          ],
        },
        Object {
          "flex": 0,
          "height": 143,
          "width": 143,
        },
      ]
    }
    width={143}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={64}
        stroke={4278365935}
        strokeWidth={15}
      />
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={64}
        stroke={4293651954}
        strokeDasharray={
          Array [
            "402.1238596594935",
            "402.1238596594935",
          ]
        }
        strokeDashoffset={-120.63715789784804}
        strokeWidth={16}
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4278216396}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "18px",
            "letterSpacing": 3,
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        rotate={Array []}
        stroke={4278216396}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            71.5,
          ]
        }
      >
        <RNSVGTSpan
          content="31,500"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4287667104}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "12px",
            "fontWeight": "bold",
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            86.5,
          ]
        }
      >
        <RNSVGTSpan
          content="OF 34,999"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`ProgressRing should render with specific color 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    bbHeight={143}
    bbWidth={143}
    focusable={false}
    height={143}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "transform": Array [
            Object {
              "rotateZ": "270deg",
            },
          ],
        },
        Object {
          "flex": 0,
          "height": 143,
          "width": 143,
        },
      ]
    }
    width={143}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={64}
        stroke={4278365935}
        strokeWidth={15}
      />
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={64}
        stroke={4293651954}
        strokeDasharray={
          Array [
            "402.1238596594935",
            "402.1238596594935",
          ]
        }
        strokeDashoffset={-120.63715789784804}
        strokeWidth={16}
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4278216396}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "18px",
            "letterSpacing": 3,
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        rotate={Array []}
        stroke={4278216396}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            71.5,
          ]
        }
      >
        <RNSVGTSpan
          content="31,500"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4287667104}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "12px",
            "fontWeight": "bold",
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            86.5,
          ]
        }
      >
        <RNSVGTSpan
          content="OF 34,999"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`ProgressRing should render with specific strokeWidth 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    bbHeight={143}
    bbWidth={143}
    focusable={false}
    height={143}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "transform": Array [
            Object {
              "rotateZ": "270deg",
            },
          ],
        },
        Object {
          "flex": 0,
          "height": 143,
          "width": 143,
        },
      ]
    }
    width={143}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={64}
        stroke={4278365935}
        strokeWidth={15}
      />
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={64}
        stroke={4293651954}
        strokeDasharray={
          Array [
            "402.1238596594935",
            "402.1238596594935",
          ]
        }
        strokeDashoffset={-120.63715789784804}
        strokeWidth={16}
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4278216396}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "18px",
            "letterSpacing": 3,
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        rotate={Array []}
        stroke={4278216396}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            71.5,
          ]
        }
      >
        <RNSVGTSpan
          content="31,500"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4287667104}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "12px",
            "fontWeight": "bold",
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            86.5,
          ]
        }
      >
        <RNSVGTSpan
          content="OF 34,999"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`ProgressRing should render with specific width  1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    bbHeight={143}
    bbWidth={143}
    focusable={false}
    height={143}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "transform": Array [
            Object {
              "rotateZ": "270deg",
            },
          ],
        },
        Object {
          "flex": 0,
          "height": 143,
          "width": 143,
        },
      ]
    }
    width={143}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={64}
        stroke={4278365935}
        strokeWidth={15}
      />
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={64}
        stroke={4293651954}
        strokeDasharray={
          Array [
            "402.1238596594935",
            "402.1238596594935",
          ]
        }
        strokeDashoffset={-120.63715789784804}
        strokeWidth={16}
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4278216396}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "18px",
            "letterSpacing": 3,
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        rotate={Array []}
        stroke={4278216396}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            71.5,
          ]
        }
      >
        <RNSVGTSpan
          content="31,500"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4287667104}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "12px",
            "fontWeight": "bold",
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            86.5,
          ]
        }
      >
        <RNSVGTSpan
          content="OF 34,999"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`ProgressRing should render with zero progress 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      undefined,
    ]
  }
>
  <RNSVGSvgView
    bbHeight={143}
    bbWidth={143}
    focusable={false}
    height={143}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "transform": Array [
            Object {
              "rotateZ": "270deg",
            },
          ],
        },
        Object {
          "flex": 0,
          "height": 143,
          "width": 143,
        },
      ]
    }
    width={143}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        r={64}
        stroke={4278365935}
        strokeWidth={15}
      />
      <RNSVGCircle
        cx={71.5}
        cy={71.5}
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={64}
        stroke={4293651954}
        strokeDasharray={
          Array [
            "402.1238596594935",
            "402.1238596594935",
          ]
        }
        strokeDashoffset={null}
        strokeWidth={16}
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4278216396}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "18px",
            "letterSpacing": 3,
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
            "stroke",
          ]
        }
        rotate={Array []}
        stroke={4278216396}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            71.5,
          ]
        }
      >
        <RNSVGTSpan
          content="31,500"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4287667104}
        font={
          Object {
            "fontFamily": "SFProDisplay-Bold",
            "fontSize": "12px",
            "fontWeight": "bold",
            "textAnchor": "middle",
          }
        }
        matrix={
          Array [
            6.123233995736766e-17,
            1,
            -1,
            6.123233995736766e-17,
            143,
            0,
          ]
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            71.5,
          ]
        }
        y={
          Array [
            86.5,
          ]
        }
      >
        <RNSVGTSpan
          content="OF 34,999"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
      <RNSVGCircle
        cx={135.5}
        cy={71.5}
        fill={4278365935}
        propList={
          Array [
            "fill",
          ]
        }
        r={7.5}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
